kobo/apps/accounts/adapter.py

```diff
@@ -1,10 +1,52 @@
 from allauth.account.adapter import DefaultAccountAdapter
 from allauth.account.forms import SignupForm
 from constance import config
+from django.conf import settings
+from django.contrib.auth import REDIRECT_FIELD_NAME
+from django.contrib.sites.shortcuts import get_current_site
 from django.db import transaction
+from django.shortcuts import resolve_url
+from django.template.response import TemplateResponse
+from trench.utils import get_mfa_model, user_token_generator
+
+from .mfa.forms import MfaTokenForm
+from .mfa.views import MfaTokenView
 
 
 class AccountAdapter(DefaultAccountAdapter):
+
+    def pre_login(self, request, user, **kwargs):
+
+        auth_method = (
+            get_mfa_model()
+            .objects.filter(is_active=True, user=user)
+            .first()
+        )
+        if auth_method:
+            ephemeral_token_cache = user_token_generator.make_token(user)
+            mfa_token_form = MfaTokenForm(
+                initial={'ephemeral_token': ephemeral_token_cache}
+            )
+            next_url = kwargs.get(
+                'redirect_url', resolve_url(settings.LOGIN_REDIRECT_URL)
+            )
+            current_site = get_current_site(request)
+            context = {
+                REDIRECT_FIELD_NAME: next_url,
+                'site': current_site,
+                'site_name': current_site.name,
+                'view': MfaTokenView,
+                'form': mfa_token_form,
+            }
+
+            return TemplateResponse(
+                request=request,
+                template='mfa_token.html',
+                context=context,
+            )
+
+        return super().pre_login(request, user, **kwargs)
```

Comment:
In any situation where you have a pattern like:
```python
if condition:
    # many lines of code
    return a

return b
```
Please, please rewrite this as:
```python
if not condition:
    return b  # i'll call this "bailing out early" for discussion's sake

# many lines of code
return a
```

This particular example, where `# many lines of code` is only 21 lines, is not too hard to read. Imagine, though, that there's enough code that it doesn't fit on a single screen, or that there are multiple nested conditionals or loops: it starts getting hard to keep track of the context while reading.

I hate to say this, because it's crudely reductionist and risks encouraging dogmatic behavior, but try to minimize amount of nesting/indentation, and where it's necessary (as of course it often is), try to minimize the number of lines inside the indented block. This applies even to logic that doesn't match the "bail out early" kind of pattern described above; an example  from earlier today of deeply-nested and large conditionals that are challenging is [this part of kobo-install](https://github.com/kobotoolbox/kobo-install/blob/e6274462282733cf92d588b3f8366259707a5758/helpers/config.py#L1578-L1731). In particular, the `            else:` (indented 12 spaces) on line 1709 makes me scroll up 124 lines (multiple screens in my editor and on GitHub) to find its matching `if`. There may very well _not_ be a way to reduce indentation and make the code easier to read simply by reordering the logic, but in that case, the logic should be separated into different functions.